Add SSL_OP_LEGACY_EC_POINT_FORMATS and change default to allow only uncompressed point format

pimterry opened the pull request:
The internal fields and implementation for configuration of this parameter already existed, but was not exposed. This change adds simple setters to allow configuration of this field.

This fixes https://github.com/openssl/openssl/issues/12644 (already closed but never actually resolved).

This is useful for configurability generally, but primarily because OpenSSL's default for EC point formats is notably different from other TLS clients' (both Firefox & Chrome appear to always send 'uncompressed' as the only option). This isn't a problem functionally, but makes OpenSSL TLS client hellos extremely fingerprintable. Being able to change the EC point formats has been a repeated request elsewhere: https://github.com/nodejs/help/issues/2905, https://stackoverflow.com/questions/42317175/how-to-pass-ec-point-format-list-using-ssl-ctx, https://github.com/openssl/openssl/issues/8067, https://github.com/openssl/openssl/discussions/23443.

With this API it's possible for downstream users to change this if required, without making any changes to existing default behaviour. In future it might be interesting to change the default entirely to align with the rest of the ecosystem, but this seems like an easier first step in this direction that will be useful regardless.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

pimterry commented:
The remaining coding style issue here was from a neighbouring line, that isn't actually touched by this PR. I've added a separate commit to clear that up regardless so everything (hopefully) passes.

pimterry commented:
MacOS & no-ml-dsa CI jobs (respectively a failing QUIC test and a job that's now stalled for 2h+) seem like unrelated issues, since they both passed on the previous commit & the last change is just indentation fixes for the linter. I suspect they'll pass fine if somebody can rerun those.

With that I think this is ready for review. Happy to make changes or discuss further if anybody has feedback.

mattcaswell commented on CHANGES.md:
```diff
@@ -252,6 +252,11 @@ OpenSSL 3.5
+ * Add SSL_CTX_set_ec_point_formats() and SSL_set_ec_point_formats() to allow
+   configuration of the EC point formats offered in TLS client hellos.
+
+   *Tim Perry*
```
This PR has just missed the deadline for inclusion in 3.5 - so this CHANGES.md entry will have to be moved to the 3.6 section (once its there after the 3.5 alpha is released next week).

mattcaswell commented on doc/man3/SSL_CTX_set_ec_point_formats.pod:
```diff
@@ -0,0 +1,50 @@
+ #include <openssl/ssl.h>
+
+ int SSL_CTX_set_ec_point_formats(SSL_CTX *ctx, const unsigned char *formats,
+                                  unsigned int formats_len);
```
I would probably use `size_t` for this length

mattcaswell commented on doc/man3/SSL_CTX_set_ec_point_formats.pod:
```diff
@@ -0,0 +1,50 @@
+
+ int SSL_CTX_set_ec_point_formats(SSL_CTX *ctx, const unsigned char *formats,
+                                  unsigned int formats_len);
+ int SSL_set_ec_point_formats(SSL *ssl, const unsigned char *formats,
```
Probably we should call both of the these functions `set1` rather than just `set` since the memory ownership does not transfer.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -3473,6 +3473,68 @@ char *SSL_get_shared_ciphers(const SSL *s, char *buf, int size)
+    for (i = 0; i < formats_len; i++) {
+        if (formats[i] < TLSEXT_ECPOINTFORMAT_first ||
+            formats[i] > TLSEXT_ECPOINTFORMAT_last)
+            return 0;
```
Calling `ERR_raise` here to give the caller a clue as to what went wrong would not be a bad idea.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -3473,6 +3473,68 @@ char *SSL_get_shared_ciphers(const SSL *s, char *buf, int size)
+    for (i = 0; i < formats_len; i++) {
+        if (formats[i] < TLSEXT_ECPOINTFORMAT_first ||
+            formats[i] > TLSEXT_ECPOINTFORMAT_last)
+            return 0;
```
ERR_raise

mattcaswell commented on test/sslapitest.c:
```diff
@@ -10985,6 +10985,60 @@ static int test_set_alpn(void)
+
+    testresult = 1;
+
+end:
```
I might not be a bad idea to run an openssl connection with a non-default ec point formats setting to confirm this completes successfully.

mattcaswell commented on util/libssl.num:
```diff
@@ -606,3 +606,5 @@ SSL_CTX_set_domain_flags                ?	3_5_0	EXIST::FUNCTION:
 SSL_CTX_get_domain_flags                ?	3_5_0	EXIST::FUNCTION:
 SSL_get_domain_flags                    ?	3_5_0	EXIST::FUNCTION:
 SSL_CTX_set_new_pending_conn_cb         ?	3_5_0	EXIST::FUNCTION:
+SSL_CTX_set_ec_point_formats            ?	3_5_0	EXIST::FUNCTION:
```
Again - this is going to end up being 3.6

davidben commented:
It's worth noting this entire feature has been deprecated already. TLS 1.3 stops using this extension, and the TLS 1.2 ECC machinery was updated in RFC 8422 to deprecated all but the uncompressed form:
https://www.rfc-editor.org/rfc/rfc8422.html#section-5.1.2

Given that, making this configurable seems... bizarre. Skimming the use cases, it seems they all are to make OpenSSL only send uncompressed... which is that RFC 8422 mandates anyway. Perhaps a simpler and more standards-conformant change would be to align OpenSSL with RFC 8422.

pimterry commented:
Changing the default would be fine by me too, and that would align with TLS implementations elsewhere.

My main concern was that that would be more disruptive, since it has a far wider immediate effect than just exposing the existing fields to reconfigure this, but for me personally either option would work nicely.

pimterry commented:
@mattcaswell I think I've resolved all review comments here with those latest changes.

@davidben I'll open a separate PR with the alternative approach of changing the defaults. There's tradeoffs but from my point of view either option would be an excellent improvement.

ghen2 commented:
If this has to be configurable at all, a compile-time option would perhaps be more appropriate?

pimterry commented:
@ghen2 @davidben Personally, my preference has changed towards supporting runtime configuration of this for now _and_ changing the default. Would that be reasonable?

While these formats are indeed deprecated, and uncompressed-only would match modern browser behaviour so is likely safe & broadly compatible at least on the web, these formats are still supported elsewhere, and may well be used. It's hard to know the impact removing support for them completely and without alternative might cause. Additionally, the use cases above all relate to either TLS fingerprinting or edge case testing, and in both scenarios, having more flexibility to configure this dynamically is useful, for as long as these are present (even if just advertised in handshakes) elsewhere in the ecosystem.

Changing the default and add a runtime API to configure the setting seems like a good first step on that basis. That way we actively move towards standard behaviour and away from non-deprecated features by default, but if there are issues then people can easily opt-out of the change as required for some individual connections, or to configure formats freely to test edge cases or modify TLS fingerprints.

I think runtime configuration is better than compile time as downstream distributions are unlikely to know whether this is an important feature for users either, and there's no significant downside to having these formats optionally available if required (i.e. it doesn't have any security issues AFAICT, it's deprecated and removed from TLSv1.3 just because it's not often used).

Once that's been released, it'll be easier to tell if anybody is actually using it, e.g. by surveying code on GitHub, seeing if there are any reports of issues with the changed default, etc.

This is basically the middle-ground of this PR & #27065. Would people be happy with that approach?

mattcaswell commented:
> This is basically the middle-ground of this PR & https://github.com/openssl/openssl/pull/27065. Would people be happy with that approach?

The seems like a good compromise position to me.

t8m commented:
> @ghen2 @davidben Personally, my preference has changed towards supporting runtime configuration of this for now _and_ changing the default. Would that be reasonable?
> 
> While these formats are indeed deprecated, and uncompressed-only would match modern browser behaviour so is likely safe & broadly compatible at least on the web, these formats are still supported elsewhere, and may well be used. It's hard to know the impact removing support for them completely and without alternative might cause. Additionally, the use cases above all relate to either TLS fingerprinting or edge case testing, and in both scenarios, having more flexibility to configure this dynamically is useful, for as long as these are present (even if just advertised in handshakes) elsewhere in the ecosystem.
> 
> Changing the default and add a runtime API to configure the setting seems like a good first step on that basis. That way we actively move towards standard behaviour and away from non-deprecated features by default, but if there are issues then people can easily opt-out of the change as required for some individual connections, or to configure formats freely to test edge cases or modify TLS fingerprints.
> 
> I think runtime configuration is better than compile time as downstream distributions are unlikely to know whether this is an important feature for users either, and there's no significant downside to having these formats optionally available if required (i.e. it doesn't have any security issues AFAICT, it's deprecated and removed from TLSv1.3 just because it's not often used).
> 
> Once that's been released, it'll be easier to tell if anybody is actually using it, e.g. by surveying code on GitHub, seeing if there are any reports of issues with the changed default, etc.
> 
> This is basically the middle-ground of this PR & #27065. Would people be happy with that approach?

+1 to this. IMO the runtime configuration could be done via a new SSL_OP option value.

pimterry commented:
I've updated this PR to include the change from #27065 to modify the default EC point formats, in addition to making them configurable. I think this is ready for another review.

As part of this, I've worked out why the three tests failed in that other PR and fixed those here. There was actually just one test really broken by the changed default, and I've fixed that by using the new APIs to re-enabling the required point format in the config for that test, which fixes it nicely. I'll close the other PR now.

> IMO the runtime configuration could be done via a new SSL_OP option value.

That said, I _haven't_ done this. Is there a reason that approach would be better? This would imply either only supporting a fixed set of configurations (e.g. the new & old defaults) or separate flags for each of the 3 supported point formats. Personally I'm find with the current API here, and the hope is that it'll be removed in the not too distant future anyway, but open to changes if there's a strong preference for another direction.

ghen2 commented:
This works nicely, thanks!

t8m commented:
> That said, I _haven't_ done this. Is there a reason that approach would be better? This would imply either only supporting a fixed set of configurations (e.g. the new & old defaults) or separate flags for each of the 3 supported point formats. Personally I'm find with the current API here, and the hope is that it'll be removed in the not too distant future anyway, but open to changes if there's a strong preference for another direction.

My reason is that we basically need to support just 2 configurations - the existing one and the uncompressed point format only. With that it would be just one SSL_OP that would re-enable the existing configuration. Problem with adding an extra API call is that it cannot really be removed any time soon, once it is added. It has to be properly deprecated and then removed in a subsequent major version. And of course it would be extremely weird to add it and deprecate it at once so the removal could happen in 5.0 at the earliest I assume. SSL_OP is much more lightweight and can be just changed to an ignored flag if we will decide to do so in that not-so distant future.

pimterry commented:
> My reason is that we basically need to support just 2 configurations

This all makes sense. I've added a commit to add a `SSL_OP_LEGACY_EC_POINT_FORMATS` flag instead, and corresponding options.

The default is now 'uncompressed', and with this new change the only configurability is now switching between 'all 3 formats' (previous default) and just 'uncompressed-only' (new default), configured with a flag to make it easy to quietly drop support for the previous default entirely in future, if we can.

Does that work for everybody?

This is force-pushed with a rebase, since there was a conflict with master on CHANGES.

t8m commented on ssl/t1_lib.c:
```diff
@@ -1767,13 +1771,16 @@ void tls1_get_formatlist(SSL_CONNECTION *s, const unsigned char **pformats,
         *num_formats = s->ext.ecpointformats_len;
-    } else {
-        *pformats = ecformats_default;
+    } else if (s->options & SSL_OP_LEGACY_EC_POINT_FORMATS) {
```
Style nit: `s->options & SSL_OP_LEGACY_EC_POINT_FORMATS != 0`

t8m commented on test/recipes/75-test_quicapi_data/ssltraceref.txt:
```diff
@@ -19,10 +19,8 @@ Header:
           001e - 04 80 08 00 00 07 04 80-08 00 00 08 02 40 64   .............@d
           002d - 09 02 40 64                                    ..@d
-        extension_type=ec_point_formats(11), length=4
+        extension_type=ec_point_formats(11), length=2
           uncompressed (0)
-          ansiX962_compressed_prime (1)
-          ansiX962_compressed_char2 (2)
```
I assume that the same change needs to be applied to the zlib variant of ssltraceref.txt. Unfortunately we test that one in on-merge CI only.

t8m reviewed (changes requested):
Good work! Only two nits.

t8m commented on ssl/t1_lib.c:
```diff
@@ -1771,7 +1771,7 @@ void tls1_get_formatlist(SSL_CONNECTION *s, const unsigned char **pformats,
         *pformats = s->ext.ecpointformats;
         *num_formats = s->ext.ecpointformats_len;
-    } else if (s->options & SSL_OP_LEGACY_EC_POINT_FORMATS) {
+    } else if (s->options & SSL_OP_LEGACY_EC_POINT_FORMATS != 0) {
```
Oh, sorry for incorrect suggestion. It should be `(s->options & SSL_OP_LEGACY_EC_POINT_FORMATS) != 0`

pimterry replied:
Yep, now fixed, thanks. Classic example of being too confident in a tiny change to bother rebuilding locally :laughing:

pimterry commented:
I've just forced pushed a rebase here to resolve the CHANGES.md conflicts again. @t8m approved this before, @mattcaswell did you want to review it again?

If there's no blockers to this, it would be great to get it merged so it doesn't start to conflict more significantly with other changes. Is there anything else required to make that happen?

mattcaswell commented on include/openssl/ssl.h.in:
```diff
@@ -423,8 +423,8 @@ typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
-
-#define SSL_OP_PREFER_NO_DHE_KEX                         SSL_OP_BIT(35)
+# define SSL_OP_PREFER_NO_DHE_KEX                        SSL_OP_BIT(35)
+# define SSL_OP_LEGACY_EC_POINT_FORMATS                  SSL_OP_BIT(36)
```
Please can you update doc/man3/SSL_CTX_set_options.pod to describe this new option.

pimterry replied:
Good point, done :+1:

mattcaswell reviewed (approved):
@t8m please reconfirm

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented:
Pushed to master. Thank you for your contribution.

t8m commented:
Hm... the squashed commit message is misleading but it was merged already so I suppose too late to fix it.
